File: front/discount.js
//Discount Library

const DiscountLib = {

    calculate: function(cartItems) {
        let subtotal = 0;
        let totalQuantity = 0;

        // get total price, quantity
        cartItems.forEach(item => {
            const price = parseFloat(item.price);
            const quantity = parseInt(item.quantity);
            
            subtotal += price * quantity;
            totalQuantity += quantity;
        });

        // check condition for discount
        let discountAmount = 0;
        let discountType = "No Discount";

         // buy 3+ 10% discount
        if (totalQuantity >= 3) {
            discountAmount = subtotal * 0.10;
            discountType = "Discount (10%)";
        }

        // result
        return {
            subtotal: subtotal,
            discountAmount: discountAmount,
            finalTotal: subtotal - discountAmount,
            discountType: discountType
        };
    }
};

File: lib/tasks/discount.rb
module Discount
  # discount library with money rails

  def self.calculate(cart_items)
    subtotal = cart_items.sum { |item| item.price * item.quantity }
    total_quantity = cart_items.sum(&:quantity)

    discount_amount = Money.new(0, "EUR")
    discount_type = "No Discount"

    # buy 3+ 10% discount
    if total_quantity >= 3
      discount_amount = subtotal * 0.10
      discount_type = "Bulk Discount (10%)"
    end

    {
      subtotal: subtotal,
      discount_amount: discount_amount,
      final_total: subtotal - discount_amount,
      discount_type: discount_type
    }
  end
end
